src/libwalletqt/PassphraseHelper.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2014-2020, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "PassphraseHelper.h"
+#include <QDebug>
+
+Monero::optional<std::string> PassphraseHelper::onDevicePassphraseRequest(bool & on_device)
+{
+    qDebug() << __FUNCTION__;
+    onWalletPassphraseNeeded(on_device);
+
+    if (m_passphrase_abort)
+    {
+        throw std::runtime_error("Passphrase entry abort");
+    }
+
+    on_device = m_passphrase_on_device;
+    if (!on_device) {
+        auto tmpPass = m_passphrase.toStdString();
+        m_passphrase = QString();
+        return Monero::optional<std::string>(tmpPass);
+    } else {
+        return Monero::optional<std::string>();
+    }
```

Comment:
Looks like this should be guarded by the mutex as well, no?

Reply:
After the `onWalletPassphraseNeeded` returns the passphrase entry is finished and we just return the value and set `on_device`, I think the mutex is not needed. 
I use mutex in the PassphraseHelper just to guard state when using `QWaitCondition` to make sure it works properly.

Reply:
`onDevicePassphraseRequest` and `onPassphraseEntered` are concurrent. Have a good chance to corrupt the state then.

Reply:
I think I still don't see the scenario that could cause the corruption. 
The call flow goes like 

- `thread-wallet-api`: `onDevicePassphraseRequest` -> `onWalletPassphraseNeeded` mutex & cond wait for UI, wait
- `thread-gui`: UI calls `onPassphraseEntered`, fills in data, wakes condition variable which causes unblocking of the `thread-wallet-api`
- `thread-wallet-api`: continues after line `onWalletPassphraseNeeded`, returns eventually. 

For a race to happen GUI would have to submit another `onPassphraseEntered` somehow out of order. 

To be extra safe we could add 

```cpp
    m_mutex_pass.lock();
    auto cleanup = qScopeGuard([&] { m_mutex_pass.unlock; });
    if (m_passphrase_abort)
    {
        throw std::runtime_error("Passphrase entry abort");
    }

    on_device = m_passphrase_on_device;
    if (!on_device) {
        auto tmpPass = m_passphrase.toStdString();
        m_passphrase = QString();
        return Monero::optional<std::string>(tmpPass);
    } else {
        return Monero::optional<std::string>();
    }
```

What do you think?

Reply:
Looks better, I would use `QMutexLocker` instead of `qScopeGuard` though.

What do you think about dropping `onWalletPassphraseNeeded` (moving it to `onDevicePassphraseRequest`)?

Reply:
Done, pls check

Reply:
Seems fine to me now